### ISSUE TITLE
clang-tidy misc-include-cleaner does not detect `chrono` header

### ISSUE THREAD

**Issue by andrea-cassioli-maersk** [llvm/llvm-project#98122]
I have noticed that when running with `misc-include-cleaner` option, clang-tidy seems not to detect that the `chrono` header is provided and used.

In one of my project file you can find

```
#include <algorithm>
#include <chrono>
#include <functional>
#include <set>
#include <string>

// some other code

 const auto time_residual = rotation_time - std::chrono::round<std::chrono::weeks>(t);
  constexpr std::chrono::minutes kThreshold{30};
  if (std::chrono::abs(time_residual) > kThreshold) {
    results << ErrorMessage("FOO");
  }

// and more code goes here
```

Running clang-tidy I get

```
/Users/<my-path>/foo.cpp:241:68: error: no header providing "std::chrono::round" is directly included [misc-include-cleaner,-warnings-as-errors]
    2 |   const auto time_residual = rotation_time - std::chrono::round<std::chrono::weeks>(t);
      |   

```

and the header `#include <__chrono/duration.h>` is added to the file.

Notice it is not only `std::chrono::round` that is not found, but `std::chrono::weeks` as well. 

I am running on 

```
➜  $ clang-tidy --version
Homebrew LLVM version 18.1.8
  Optimized build.
```

**Comment by firewave** [comment]
Related to #68213 which shows the same issue using `libstdc++`.

**Comment by mordante** [comment]
Is there anything libc++ can do or is this a clang-include-cleaner issue? Based on that it also happens with libstdc++ I expect it's a clang-include-cleaner issue.

**Comment by firewave** [comment]
I checked some fixed issues and based on 586e497f7985ddb0ca481b45f82e202d3b6f23ea I assume it is caused by missing mappings in `clang/lib/Tooling/Inclusions/Stdlib/StdSymbolMap.inc`.

**Comment by firewave** [comment]
Actually it looks like a shortcoming of the script which generates the mapping. Using `std::chrono` as example it seems it omits symbols which have a parameter specified in the documentation at https://en.cppreference.com/w/cpp/header/chrono (e.g. ` ceil(std::chrono::duration)`). Those include `ceil` and `round` which are the culprits in the related tickets.

CC @hokein

**Comment by mordante** [comment]
It's indeed documented as not supported in `llvm/clang/tools/include-mapping/gen_std.py`. Since this is not an issue in libc++ I'll remove the label.

**Comment by andrea-cassioli-maersk** [comment]
If it can help, we have seen this happening also for `chrono` types like `weeks`

**Comment by mordante** [comment]
Based on the comment in the script it happens to everything in the namespace `std::foo`. Where `foo` is `chrono`, `ranges`, `filesystem`, etc.

**Comment by hokein** [comment]
The `gen_std.py` script does support generating `std` [subnamespace](https://github.com/llvm/llvm-project/blob/main/clang/tools/include-mapping/gen_std.py#L239) symbols. It appears that the [FIXME](https://github.com/llvm/llvm-project/blob/main/clang/tools/include-mapping/gen_std.py#L17-L18) comment is stale.

The script parses all symbol lists from symbol index pages, e.g., https://en.cppreference.com/w/cpp/symbol_index/chrono, so only symbols listed there will be included. The current `StdSymbolMap.inc` file was generated from an older cppreference offline book, which might miss some newly-added symbols (I guess `std::chrono::round`). I think we should consider rerunning the script on a newer cppreference offline [book](https://github.com/PeterFeicht/cppreference-doc/releases/tag/v20240610).

**Comment by firewave** [comment]
According to the Wayback Machine the `chrono` page has not changed since December 4, 2021. The missing symbols have `<>()` at the end, so maybe that is causing them to be left out.

**Comment by firewave** [comment]
I took a look at the script using the previously used version of the reference.

It skips these symbols because there are variants for these functions - e.g. `std::chrono:.ceil` can take either `std::chrono::duration` or `std::chrono::time_point`.

**Comment by andrea-cassioli-maersk** [comment]
Hi, any update on this? I see also issues when using time literals like

```
uppressed 1 warnings (1 NOLINT).
/Users/andreacassioli/workspace/foo.cpp:90:43: error: no header providing "std::operator""h" is directly included [misc-include-cleaner,-warnings-as-errors]
    3 |   const utils::Hours x = 1h;
      |                            ^
```
